File: server/exec/jv/Filter.java
/*
*  Filter source code: Applies a given filter from three implementations:
*     -Blur
*     -Grayscale
*     -Edge detection
*  Copyright (C) 2019  Sergio Isaac Mercado Silvano
*
*  This program is free software: you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation, either version 3 of the License, or
*  (at your option) any later version.
*
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with this program.  If not, see <https://www.gnu.org/licenses/>
*/

//Required libraries

import java.awt.image.BufferedImage;
import java.io.File;
import javax.imageio.ImageIO;
import java.util.concurrent.RecursiveAction;
import java.util.concurrent.ForkJoinPool;

//Must extend recursive action (does not return any value)

public class Filter extends RecursiveAction{
  public static final long serialVersionUID = 10293918281249501L;
  private static final int BLUR_WINDOW = 15;
  private static final int GRAIN = 10_000;
  private int src[], dest[], width, height, start, end;
  private String filter;

  //Constructor

  public Filter(int start, int end, int src[], int dest[], int width, int height, String filter){
    this.start = start;
    this.end = end;
    this.src = src;
    this.dest = dest;
    this.width = width;
    this.height = height;
    this.filter = filter;
  }

  //Blur method

  public void blur(int row, int col){
    int border, i, j, cells;
    int tmp_row, tmp_col, pixel, dpixel;

    float r, g, b;

    border = (BLUR_WINDOW - 1)/2;
    cells = (BLUR_WINDOW * BLUR_WINDOW);

    r = 0;
    g = 0;
    b = 0;

    for(i = -border; i <= border; i++){
      for(j = -border; j <= border; j++){
        tmp_row = Math.min( Math.max(row + i, 0), height - 1);
        tmp_col = Math.min( Math.max(col + i, 0), width - 1);

        pixel = src[(tmp_row * width) + tmp_col];

        r += (float) ((pixel & 0x00ff0000) >> 16);
        g += (float) ((pixel & 0x0000ff00) >> 8);
        b += (float) ((pixel & 0x000000ff) >> 0);
      }
    }

    dpixel = (0xff000000)
                         | (((int) (r / cells)) << 16)
                         | (((int) (g / cells)) << 8)
                         | (((int) (b / cells)) << 0);

    dest[(row * width) + col] = dpixel;
  }

  //Grayscale method

  private void gray(int row, int col){
    int pixel, dpixel;

    float r, g, b, avg;

    pixel = src[(row * width) + col];

    r = (float) ((pixel & 0x00ff0000) >> 16);
    g = (float) ((pixel & 0x0000ff00) >> 8);
    b = (float) ((pixel & 0x000000ff) >> 0);

    avg = (r + g + b)/3;

    dpixel = (0xff000000)
                         | ((int) avg << 16)
                         | ((int) avg << 8)
                         | ((int) avg << 0);

    dest[(row * width) + col] = dpixel;
  }

  //Edge detection method

  public void edge(int row, int col){
    int topPixel, lowPixel, dpixel;
    int tmp_row;

    float rH, gH, bH, avgH;
    float rL, gL, bL, avgL;

    tmp_row = Math.min( Math.max(row + 1, 0), height - 1);

    topPixel = src[(row * width) + col];
    lowPixel = src[(tmp_row * width) + col];

    rH = (float) ((topPixel & 0x00ff0000) >> 16);
    gH = (float) ((topPixel & 0x0000ff00) >> 8);
    bH = (float) ((topPixel & 0x000000ff) >> 0);

    rL = (float) ((lowPixel & 0x00ff0000) >> 16);
    gL = (float) ((lowPixel & 0x0000ff00) >> 8);
    bL = (float) ((lowPixel & 0x000000ff) >> 0);

    avgH = (rH + gH + bH)/3;
    avgL = (rL + gL + bL)/3;

    if(0.65 >= Math.abs(avgH - avgL) &&
       0.70 >= Math.abs(avgH - avgL)){
      dpixel = (0xffffffff);
    }else{
      dpixel = (0xff000000);
    }

    dest[(row * width) + col] = dpixel;
  }


  //Serial filtering

  protected void computeDirectly(){
    int index;
    int row, col;

    for(index = start; index < end; index++){
      row = index / width;
      col = index % width;

      //Check which filter to apply on each iteration

      switch(filter){
        case "blur":
          blur(row, col);
          break;
        case "edge":
          edge(row, col);
          break;
        case "gray":
          gray(row, col);
          break;
        default:
          break;
      }
    }
  }

  //RecursiveAction required method (compute())

  @Override
  protected void compute(){

    //Check if problem can be solved serially
    if((end - start) <= GRAIN){
      computeDirectly();
    }else{

      //Split the size in half on each iteration

      int mid = start + ((end - start) / 2);

      //Let all threads work on upper and lower half on each iteration
      invokeAll(new Filter(start, mid, src, dest, width, height, filter),
                new Filter(mid, end, src, dest, width, height, filter));
    }
  }

  public static void main(String[] args) throws Exception{
    ForkJoinPool pool;

    //Command parameters

    final String srcName = args[0];
    final String filterT = args[1];
    final String destName = args[2];

    File srcFile = new File(srcName);

    final BufferedImage source = ImageIO.read(srcFile);

    //Obtain image parameters

    int w = source.getWidth();
    int h = source.getHeight();

    //Transform image into array
    int src[] = source.getRGB(0, 0, w, h, null, 0, w);

    //Prepare destination array of same size as source
    int dest[] = new int[src.length];

    //Begin Thread Pool work

    pool = new ForkJoinPool(Runtime.getRuntime().availableProcessors());
    pool.invoke(new Filter(0, w * h, src, dest, w, h, filterT));

    //End Thread Pool work

    //Transform array into image
    final BufferedImage destination = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
    destination.setRGB(0, 0, w, h, dest, 0, w);

    //Write image file into "img" directory with new filename
    File output = new File("img/" + destName);
    ImageIO.write(destination, "png", output);
  }
}

File: server/exec/omp/filter.c
/*
*  Filter source code: Applies a given filter from three implementations:
*     -Blur
*     -Grayscale
*     -Edge detection
*  Copyright (C) 2019  Sergio Isaac Mercado Silvano
*
*  This program is free software: you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation, either version 3 of the License, or
*  (at your option) any later version.
*
*  This program is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  You should have received a copy of the GNU General Public License
*  along with this program.  If not, see <https://www.gnu.org/licenses/>
*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <opencv/highgui.h>
#include <omp.h>

#define BLUR_WINDOW 15

typedef enum color {BLUE, GREEN, RED} Color;

void blur(IplImage *src, IplImage *dest, int ren, int col){
  int side_pixels, i, j, cells;
  int tmp_ren, tmp_col, step;
  float r, g, b;

  side_pixels = (BLUR_WINDOW - 1)/2;
  cells = (BLUR_WINDOW * BLUR_WINDOW);
  step = src->widthStep/sizeof(uchar);

  r = 0;
  g = 0;
  b = 0;

  for(i = -side_pixels; i <= side_pixels; i++){
    for(j = -side_pixels; j <= side_pixels; j++){
      tmp_ren = MIN(MAX(ren + i, 0), src->height - 1);
      tmp_col = MIN(MAX(col + j, 0), src->width - 1);

      r += (float) src->imageData[(tmp_ren * step) +
                                  (tmp_col * src->nChannels) + RED];

      g += (float) src->imageData[(tmp_ren * step) +
                                  (tmp_col * src->nChannels) + GREEN];

      b += (float) src->imageData[(tmp_ren * step) +
                                  (tmp_col * src->nChannels) + BLUE];
    }
  }

  dest -> imageData[(ren * step) + (col * dest->nChannels) + RED] = (unsigned char) (r / cells);
  dest -> imageData[(ren * step) + (col * dest->nChannels) + GREEN] = (unsigned char) (g / cells);
  dest -> imageData[(ren * step) + (col * dest->nChannels) + BLUE] = (unsigned char) (b / cells);
}

void gray(IplImage *src, IplImage *dest, int row, int col){
  float r, g, b, avg;
  int step;

  step = src->widthStep/sizeof(uchar);

  r = 0; g = 0; b = 0;

  r = (float) src->imageData[(row * step) + (col * src->nChannels) + RED];
  g = (float) src->imageData[(row * step) + (col * src->nChannels) + GREEN];
  b = (float) src->imageData[(row * step) + (col * src->nChannels) + BLUE];

  avg = (r + g + b)/3.0;

  dest -> imageData[(row * step) + (col * dest->nChannels) + RED] =  (unsigned char) (avg);
  dest -> imageData[(row * step) + (col * dest->nChannels) + GREEN] = (unsigned char) (avg);
  dest -> imageData[(row * step) + (col * dest->nChannels) + BLUE] = (unsigned char) (avg);
}

void edge(IplImage *src, IplImage *dest, int row, int col){
  int tmp_row, step;

  step = src->widthStep/sizeof(uchar);

  float rH, gH, bH, avgH;
  float rL, gL, bL, avgL;

  rH = 0; gH = 0; bH = 0;
  rL = 0; gL = 0; bL = 0;

  tmp_row = MIN(MAX(row + 1, 0), src->height - 1);

  rH = (float) src->imageData[(row * step) + (col * src->nChannels) + RED];
  gH = (float) src->imageData[(row * step) + (col * src->nChannels) + GREEN];
  bH = (float) src->imageData[(row * step) + (col * src->nChannels) + BLUE];

  rL = (float) src->imageData[(tmp_row * step) + (col * src->nChannels) + RED];
  gL = (float) src->imageData[(tmp_row * step) + (col * src->nChannels) + GREEN];
  bL = (float) src->imageData[(tmp_row * step) + (col * src->nChannels) + BLUE];

  avgH = (rH + gH + bH)/3.0;
  avgL = (rL + gL + bL)/3.0;

  if((0.65 >= fabs(avgH - avgL)) && (0.70 >= fabs(avgH - avgL))){
    dest -> imageData[(row * step) + (col * dest->nChannels) + RED] = (unsigned char) (0xFF);
    dest -> imageData[(row * step) + (col * dest->nChannels) + GREEN] = (unsigned char) (0xFF);
    dest -> imageData[(row * step) + (col * dest->nChannels) + BLUE] = (unsigned char) (0xFF);
  }else{
    dest -> imageData[(row * step) + (col * dest->nChannels) + RED] = (unsigned char) (0);
    dest -> imageData[(row * step) + (col * dest->nChannels) + GREEN] = (unsigned char) (0);
    dest -> imageData[(row * step) + (col * dest->nChannels) + BLUE] = (unsigned char) (0);
  }
}

void apply(IplImage *src, IplImage *dest, char * flt){
  int index, size;
  int ren, col;

  size = src->width * src->height;

  if(strcmp(flt,"blur") == 0){
    #pragma omp parallel for shared(src, dest, size) private(ren, col)
    for(index = 0; index < size; index++){
      ren = index / src->width;
      col = index % src->width;
      blur(src,dest,ren,col);
    }
  }

  if(strcmp(flt,"gray") == 0){
    #pragma omp parallel for shared(src, dest, size) private(ren, col)
    for(index = 0; index < size; index++){
      ren = index / src->width;
      col = index % src->width;
      gray(src,dest,ren,col);
    }
  }

  if(strcmp(flt,"edge") == 0){
    #pragma omp parallel for shared(src, dest, size) private(ren, col)
    for(index = 0; index < size; index++){
      ren = index / src->width;
      col = index % src->width;
      edge(src,dest,ren,col);
    }
  }
}

int main(int argc, char *argv[]) {
  int i;
  char dest_path[256] = "img/";

  IplImage *src = cvLoadImage(argv[1], CV_LOAD_IMAGE_COLOR);
  IplImage *dest = cvCreateImage(cvSize(src->width, src->height), IPL_DEPTH_8U, 3);

  apply(src, dest, argv[2]);


  strcat(dest_path,argv[3]);

  cvSaveImage(dest_path, dest, 0);

  return 0;
}
